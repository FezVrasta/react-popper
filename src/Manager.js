// @flow
import React, { Component, type Node } from 'react';
import createContext, { type Context } from 'create-react-context';

export const ManagerContext: Context<{
  getReferenceRef?: (?HTMLElement) => void,
  referenceNode?: ?HTMLElement,
}> = createContext({});

type ManagerProps = {
  children: Node,
};
type ManagerState = {
  referenceNode?: ?HTMLElement,
};

export default class Manager extends Component<ManagerProps, ManagerState> {
  state = {};

  getReferenceRef = (referenceNode: ?HTMLElement) =>
    this.setState({ referenceNode });

  render() {
    return (
      <ManagerContext.Provider
        value={{
          referenceNode: this.state.referenceNode,
          getReferenceRef: this.getReferenceRef,
        }}
      >
        {this.props.children}
      </ManagerContext.Provider>
    );
  }
}
