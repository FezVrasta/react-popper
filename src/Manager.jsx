import { Component, createElement } from 'react'
import PropTypes from 'prop-types'

class Manager extends Component {
  static childContextTypes = process.env.NODE_ENV !== 'production'
    ? {
        popperManager: PropTypes.object.isRequired,
      }
    : {
        popperManager: () => {},
      }

  static propTypes = {
    tag: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  }

  static defaultProps = {
    tag: 'div',
  }

  getChildContext() {
    return {
      popperManager: {
        setTargetNode: this._setTargetNode,
        getTargetNode: this._getTargetNode,
      },
    }
  }

  _setTargetNode = node => {
    this._targetNode = node
  }

  _getTargetNode = () => {
    return this._targetNode
  }

  render() {
    const { tag, children, ...restProps } = this.props
    if (tag !== false) {
      return createElement(tag, restProps, children)
    } else {
      return children
    }
  }
}

export default Manager
