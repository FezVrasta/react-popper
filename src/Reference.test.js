// @flow
import React from 'react';
import { mount } from 'enzyme';

// Public API
import { Reference } from '.';

// Private API
import { ManagerContext } from './Manager';

describe('Arrow component', () => {
  it('renders the expected markup', () => {
    const wrapper = mount(
      <Reference>{({ ref }) => <div ref={ref} />}</Reference>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('consumes the ArrowNodeContext from Popper', () => {
    const getReferenceRef = jest.fn();

    // HACK: wrapping DIV needed to make Enzyme happy for now
    mount(
      <div>
        <ManagerContext.Provider
          value={{
            getReferenceRef,
            referenceNode: undefined,
          }}
        >
          <Reference>{({ ref }) => <div ref={ref} />}</Reference>
        </ManagerContext.Provider>
      </div>
    );
    expect(getReferenceRef).toHaveBeenCalled();
  });
});
