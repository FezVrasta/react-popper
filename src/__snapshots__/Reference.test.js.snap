// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow component renders the expected markup 1`] = `
<Reference>
  <InnerReference
    getReferenceRef={
      [MockFunction] {
        "calls": Array [
          Array [
            <div />,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  >
    <div />
  </InnerReference>
</Reference>
`;
